include/openthread/ip6.h

```diff
@@ -139,13 +139,17 @@ typedef struct otSockAddr
  */
 typedef struct otMessageInfo
 {
-    otIp6Address mSockAddr;            ///< The local IPv6 address.
-    otIp6Address mPeerAddr;            ///< The peer IPv6 address.
-    uint16_t     mSockPort;            ///< The local transport-layer port.
-    uint16_t     mPeerPort;            ///< The peer transport-layer port.
-    const void * mLinkInfo;            ///< A pointer to link-specific information.
-    uint8_t      mHopLimit;            ///< The IPv6 hop limit.
-    bool         mIsHostInterface : 1; ///< TRUE if packets sent/received via host interface, FALSE otherwise.
+    otIp6Address mSockAddr;    ///< The local IPv6 address.
+    otIp6Address mPeerAddr;    ///< The peer IPv6 address.
+    uint16_t     mSockPort;    ///< The local transport-layer port.
+    uint16_t     mPeerPort;    ///< The peer transport-layer port.
+    const void * mLinkInfo;    ///< A pointer to link-specific information.
+    uint8_t      mHopLimit;    ///< The IPv6 hop limit.
+                               ///< For the default value `0`, `OPENTHREAD_CONFIG_IP6_HOP_LIMIT_DEFAULT` would be used
+                               ///< when passing down to IP layer if `mApplyHopLimit0` is FALSE by default;
+                               ///< Otherwise,  hop limit 0 is applied in IP layer if `mApplyHopLimit0` is TRUE.
+    bool mIsHostInterface : 1; ///< TRUE if packets sent/received via host interface, FALSE otherwise.
+    bool mApplyHopLimit0 : 1;  ///< TRUE if hop limit 0 is expected to take effect in IP layer.
```

Comment:
I'm fine with this design, but I prefer an initializing function for this structure.

Reply:
Agree with @bukepo. An init sounds better to me.
How about changing `mHopLimit` to be `int16_t` type instead and then use `-1` (or any negative value) as use the default `OPENTHREAD_CONFIG_IP6_HOP_LIMIT_DEFAULT` in that case.

Reply:
One concern with using init function is that this breaks existing application/user code that relies on `memset()`.

Reply:
I agree that `mApplyHopLimit0` looks like a hack, at the same time as mentioned in the previous PR/Issues the requirement to set `mHopLimit` will cause a lot of issues for OT users and existing applications (though we can argue that due to incorrect usage of API). I was leaning towards reverting Init function before due to not having a clear use case of using Hop Limit 0, which i understand we have now.

Personally then I think we should either:

1. Use @bukepo approach
2. Use a little bit different variant of @librasungirl approach:
- add `mOverrideDefaultHopLimit` variable which has to be set in order to make use of `mHopLimit`
- use `mHopLimit` value only when `mOverrideDefaultHopLimit` variable is set

Reply:
Both Init() solution by @bukepo and option `2` mentioned by @LuDuda would impact developer's experience, and some existing sample code need to be revised accordingly as well. 

I don't know, but if it is agreed to allow such impact, I am fine with @bukepo 's solution.

Reply:
I think the primary consideration is whether we continue to support initialization using `memset()`. FWIW, the OpenThread APIs generally support struct initialization using `memset()`. For that reason, I would lean towards @LuDuda's suggestion of having a separate flag that indicates whether `mHopLimit` contains a valid value. Thoughts?

Reply:
I also favor the @LuDuda approach (adding `mOverrideDefaultHopLimit` which would be by default false when using `memset`).

I share the concern of @librasungirl , that such a change would impact public APIs/behavior and existing code need to be revised. 
One suggestion to make it very clear that definition/behavior is changed: We can consider renaming the `mHopLimit` name to `mUserHopLimit` (or perhaps some other name). This way, existing code (still using `mHopLimit`) would break at compile time and developers/users would detect this and revise the use.

Thoughts?

Reply:
+1 @abtink. I think that even if we break API, this will not affect current applications much since anyway it's rare to change the hop limit value.

I am also fine with changing the name of the `mOverrideDefaultHopLimit` flag.

Reply:
@abtink The internal [`MessageInfo`](https://github.com/openthread/openthread/blob/master/src/core/net/socket.hpp#L55) inherits from `otMessageInfo`, do you mean all `mHopLimit `(APIs/core) changes to `mUserHopLImit`? Or introduce one more field `mHopLimit `in internal 'MessageInfo' to be distinguished from m`UserHopLimit`?

Agree with @LuDuda  that the proposal in this PR seems a little hack. Compared with 'mOverrideDefaultHopLimit ' option,  here I want to defend the proposal in this PR a bit:

As @LuDuda mentioned, it might be rare to change the hop limit value for the developer. 
And it would be **even more rare to use hop limit 0** for the developer (so far, only required in one test defined in 1.2 Test Plan), one advantage of this PR is that it won't impact user's experience in specifying a non-zero hoplimit.

If generally hoplimit 0 won't be used by developer (which should be true generally I guess), I would prefer this solution in this PR personally. However I am open to change if the other options are preferred as a more general solution.

Thoughts?

Reply:
@librasungirl my suggestion was to rename the `mHopLimit` in `otMessageInfo` (which would then require us to change its interna/corel use as well (which I think should be mostly abstracted behind the getter we provide from `MessageInfo::GetHopLimit()`).

Thinking more about this, I agree with @librasungirl approach and argument, that the current proposal/solution in the PR has some clear advantages (it keeps the behavior same for any current user of APIs, and it works with `memset` of struct, while adding the new capability to use hop limit value of zero (which itself should not be commonly used)). I am fine with moving on with this solution.

Reply:
@abtink thanks for your clarification.

@LuDuda @jwhui thoughts?